feat(import): Import YouTube playlist from Nightbot

redbeardymcgee opened the pull request:
This commit implements the first pass at importing the playlist managed by Nightbot. In the code, I have left a few comments to make note of some undercooked solutions.

I tried doing this with simple `React.useState()` but you may find that this approach is too naive. For now, I do it this way to make it easier to output the playlist resource into a text field. It's something like a `console.log(playlist)` with extra steps.

Setting aside code style, my next step is to figure out how to convert the tracks from Nightbot into tracks for SogeBot. Do you have some pointers to where I could start gluing that together? 

In [playlist.tsx](https://github.com/sogebot/sogebot.dev/blob/0f587695c77d19210a144ccbb3058a21f43167b0/ui.admin-mui/src/routes/manage/youtube/playlist.tsx#L263) I see a handler for adding songs to the playlist, but I'm not sure if that's the right inspiration I should find (copy and paste lmao).

Also, existing SogeBot users may discover this feature and wish to import their old Nightbot playlist. There is some potential for such an import to clobber an existing playlist, I think by modifying `createdAt` or `updatedAt` values or duplicating items. I'm not sure how to mitigate that yet.

redbeardymcgee committed:
feat(import): first pass at Nighbot playlist

This commit implements the first pass at importing the playlist managed
by Nightbot. In the code, I have left a few comments to make note of
some undercooked solutions.

I tried doing this with simple `React.useState()` but you may find that
this approach is too naive. For now, I do it this way to make it easier
to output the playlist resource into a text field. It's something like
a `console.log(playlist)` with extra steps.

redbeardymcgee committed:
feat(import): improve output and correct comments

sogehige commented:
> In [playlist.tsx](https://github.com/sogebot/sogebot.dev/blob/0f587695c77d19210a144ccbb3058a21f43167b0/ui.admin-mui/src/routes/manage/youtube/playlist.tsx#L263) I see a handler for adding songs to the playlist, but I'm not sure if that's the right inspiration I should find (copy and paste lmao).

Yes, you should use https://github.com/sogebot/sogebot.dev/blob/0f587695c77d19210a144ccbb3058a21f43167b0/ui.admin-mui/src/routes/manage/youtube/playlist.tsx#L271C8-L273C10

e.g.
```typescript
const async import = () => {
  const videoIds = ['bvNZeh6f8vE']
  for (const videoId of videoIds) {
    await new Promise((resolve, reject) => getSocket('/systems/songs').emit('import.video', {
      playlist: videoId, forcedTag: 'nightbot-import'
    }, (err) => {
      if (err) {
        reject(err)
      } else {
        resolve()
      }
    }))
  };
}
```
        
it may take a while as we are fetching video data, so thats why it is one by one, you should handle reject as well as some videos might not be available anymore

redbeardymcgee committed:
feat(import): implement first playlist import

We now have a complete cycle starting from authorizing the user with
nightbot, and ending with songs imported directly into an empty sogebot
playlist.

This is still a draft. There is no error handling, but it's very easy
to receive a 429 from the Nightbot API. I didn't see any docs that
described any rate limits, but all errors should at least contain a
`message` key describing the issue. I need to investigate how to
handle errors in a way that would be idiomatically aligned with
the SogeBot source code.

redbeardymcgee committed:
feat(import): Implement basic error handling

This commit additionally makes an effort at fixing the type errors.
One may consider this effort more of an appeasement than solution.

redbeardymcgee commented:
Due to the amount of overall churn in the diffs because of my note-taking in comments, I find this code easier to look at through this [link to the file](https://github.com/sogebot/sogebot.dev/blob/888b35526a882b613c0086eb15a956f80c38828f/ui.admin-mui/src/components/Settings/Import/nightbot.tsx#L69).

I will be removing all comments as the implementation becomes more stable. Until then, they include open questions that may answer themselves, again a result of the implementation details stabilizing. If you do have some input on them, feel free to share here.

Thank you!

redbeardymcgee committed:
feat(import): less faulty error handling

The previous attempt at error handling was a farce. It did not work at
all because I didn't actually catch the errors. I tested for results,
but those results don't return when the response is a 429 error.

This version tries to use an axios interceptor to do something when
a non-success code is returned. Right now it just prints an error
object and returns undefined (or is it null?).

Some additional effort was made to declare better types for the api
response.

redbeardymcgee committed:
feat(import): implement clearer fetching

`fetchResource()` is still recursive, but pagination is heavily
simplified by incrementing the `offset` variable as an argument
to the function.

This also re-enables full playlist importing to make it possible to
encounter errors that I was unable to witness with a small playlist.

I am not confident that an axios interceptor can be applied in order
to handle errors. It seems promising, so maybe I am just doing it
wrong. However, I am ready to dispose of it, because this technique
also forces our usage of `axios.get()` to look unclear while not
providing any actual error handling. These really seem more like
transformers, or maybe a way to verify the shape of a request or
response before committing to processing it.

sogehige commented on ui.admin-mui/src/components/Settings/Import/nightbot.tsx:
```diff
@@ -65,29 +66,156 @@ const PageSettingsModulesImportNightbot: React.FC<{
+    const videos: PlaylistItemTrack[] = await fetchPlaylist();
+
+    videos.filter((track) => {
+      // TODO: Nightbot supports SoundCloud also
```
no, we are not supporting soundcloud at all

sogehige commented on ui.admin-mui/src/components/Settings/Import/nightbot.tsx:
```diff
@@ -65,29 +66,156 @@ const PageSettingsModulesImportNightbot: React.FC<{
+          },
+          (err) => {
+            if (err) {
+              // FIXME: This could potentially generate
```
you can just count number of failures and if >0 then just enqueueSnackbar something like `x video(s) failed to import` or something like that. wdyt?

redbeardymcgee committed:
feat(import): attempt to retry on 429

redbeardymcgee committed:
feat(import): remove excessive note comments

redbeardymcgee committed:
feat(import): remove unused state, fix lint errors

This also messed with the component layout a bit. Since we don't keep
the playlist in state anymore, we can start cleaning up the UI.

redbeardymcgee committed:
feat(import): improve import failure feedback

redbeardymcgee committed:
feat(import): replace interceptor with try/catch

redbeardymcgee committed:
feat(import): remove wrapper fn

redbeardymcgee commented:
This is effectively a final draft of the implementation. What remains now are stylistic or organization issues, or bugs.

### Potential Bugs
 - Type checker returns "Cannot find name 'T'. [2304]" for `fetchPlaylist`
 - That's all, obviously. :)
 - Should I even attempt retries or just report failure, log it, and bail out?

### Style
 - `fetchPlaylist()` is recursive. Is it unclear in this style?
 - 3 instances of mutation: `maxRetries`, `acc.push()`, `failCount`
 - The types may be unnecessary, or should be moved to another file maybe?
 - I should at least log somewhere instead of a snackbar or console, right?

redbeardymcgee committed:
Merge branch 'sogebot:main' into feat/import-nightbot-playlist

sogehige commented on ui.admin-mui/src/components/Settings/Import/nightbot.tsx:
```diff
@@ -65,29 +66,124 @@ const PageSettingsModulesImportNightbot: React.FC<{
+    playlist: PlaylistItem[];
+  };
+
+  const fetchPlaylist = async (acc: PlaylistItemTrack[], offset: number): Promise<T> => {
```
```suggestion
  const fetchPlaylist = async (acc: PlaylistItemTrack[] = [], offset = 0): Promise<T> => {
```

sogehige replied:
seems little bit nicer for me, but your version is ok as well, depends on you

sogehige commented on ui.admin-mui/src/components/Settings/Import/nightbot.tsx:
```diff
@@ -65,29 +66,124 @@ const PageSettingsModulesImportNightbot: React.FC<{
+  };
+
+  const importPlaylist = async () => {
+    const videos: PlaylistItemTrack[] = await fetchPlaylist([], 0);
```
```suggestion
    const videos: PlaylistItemTrack[] = await fetchPlaylist();
```

sogehige commented on ui.admin-mui/src/components/Settings/Import/nightbot.tsx:
```diff
@@ -65,29 +66,124 @@ const PageSettingsModulesImportNightbot: React.FC<{
+            </InputAdornment>,
+          }}
+        />
+        <Stack direction='row-reverse'>
```
```suggestion
        <Stack direction='row'>
```

sogehige commented on ui.admin-mui/src/components/Settings/Import/nightbot.tsx:
```diff
@@ -65,29 +66,124 @@ const PageSettingsModulesImportNightbot: React.FC<{
+          }}
+        />
+        <Stack direction='row-reverse'>
+          <Button color='primary' variant='contained' disabled={user === 'Not Authorized'} onClick={importPlaylist}>Import</Button>
```
```suggestion
          <Button color='primary' variant='contained' disabled={user === 'Not Authorized'} onClick={importPlaylist}>Import playlist</Button>
```

redbeardymcgee committed:
feat(import): Apply suggestions from code review

Co-authored-by: sogehige <sogehige@gmail.com>

redbeardymcgee commented on ui.admin-mui/src/components/Settings/Import/nightbot.tsx:
```diff
@@ -65,29 +66,124 @@ const PageSettingsModulesImportNightbot: React.FC<{
+
+      return await fetchPlaylist(acc, offset + 100);
+    } catch (error: any) {
+      if (error.response.status === 429 && maxRetries !== 0) {
```
I think the conditional branches in here should also return the accumulator. Currently i think both else branches just fall out, returning `undefined` as a default.

redbeardymcgee committed:
feat(import): return acc from error branches

redbeardymcgee commented:
I think this is in a workable state now. The feature works fine, but of course I'm worried that I messed something up. I will remain available in case someone finds a way to break my implementation, so I can try to fix it better :)

redbeardymcgee commented:
I was testing this out and it's no longer catching the 429. Have to rework the error handling.

redbeardymcgee committed:
feat(import): reimplement retry logic

Previously, the retry logic was naive in a way that made it do nothing.
While it technically did retry a failed request, it only retried a
single time, and the result of the retry was discarded.

This version actually captures error responses and retries each failed
request up to 3 times, delayed by 1 minute each. This will block the
importer from receiving the playlist tracks from Nightbot for
potentially a long time, but will also gracefully handle 429 errors.

Using a generator function may allow for processing a page of tracks
while other requests are awaiting a response. It is unlikely such an
optimization will be necessary, since this feature is expected to be
used only 1 time per user. Additionally, the number of users who will
care to import their playlist from Nightbot is probably very small.

redbeardymcgee commented:
Copying the commit message body here for visibility:

> Previously, the retry logic was naive in a way that made it do nothing.
> While it technically did retry a failed request, it only retried a
> single time, and the result of the retry was discarded.
> 
> This version actually captures error responses and retries each failed
> request up to 3 times, delayed by 1 minute each. This will block the
> importer from receiving the playlist tracks from Nightbot for
> potentially a long time, but will also gracefully handle 429 errors.
> 
> Using a generator function may allow for processing a page of tracks
> while other requests are awaiting a response. It is unlikely such an
> optimization will be necessary, since this feature is expected to be
> used only 1 time per user. Additionally, the number of users who will
> care to import their playlist from Nightbot is probably very small.

Now there is the matter of http 410 errors from YouTube when a video has been delisted. The error happens during the import step on the backend, which I have been hesitant to touch. Previously, I thought it was just being discarded by SogeBot, which is probably the best thing to do aside from logging the error. It doesn't seem correct for me to make a request to YouTube to check that the video is still available, since SogeBot will then need to make a second request for the same video ID when importing. Does that make sense?

redbeardymcgee committed:
feat(import): use consistent func definition style

redbeardymcgee committed:
feat(import): swallow 410 error

redbeardymcgee committed:
feat(import): output failure as error, not log

redbeardymcgee commented:
For now I've chosen to swallow the errors returned by the video import, since there really isn't some useful thing to do when it fails. Almost certainly it's a result of a delisted video.

If you have no objections, I think this can be considered complete. Please let me know if there's anything else I can add or change.
